Validation and serialization (#2023)
fastify/fastify: docs/Server.md

<a name="add-schema"></a>
#### addSchema

> `fastify.addSchema(schemaObj)`, adds a JSON schema to the Fastify instance. This allows you to reuse it everywhere in your application just by using the standard `$ref` keyword.<br/>

Reviewer: Do we really want to stick with this feature? Given the complexity in maintaining proper JSON Schema instances with references, wouldn't it be better to relegate everything to a passed in schema container?

Author: I'm not sure it could works because the users that use the default validator and serializer need to add a common schema, and it would not be possible without this function.
Moreover, right now we are not managing the `$ref`s, so I think it would be better from now on

But maybe I misunderstand your idea
Could you elaborate?

> To learn more, checkout the [Validation and Serialization](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md) documentation.

Reviewer: ```suggestion
To learn more, read the [Validation and Serialization](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md) documentation.
```

<a name="get-schemas"></a>
#### getSchemas

> `fastify.getSchemas()`, returns a JSON with all the schemas added with `.addSchema`. The keys of the JSON are the `$id`s of the JSON schema provided.

Reviewer: What is "a JSON"? Are you saying this method returns one big object with all schemas attached to properties?

Author: Here the case:

```
fastify.addSchema({ $id: 'A', properties: { hello: '...' } })
fastify.addSchema({ $id: 'B', properties: { hello: '...' } })

fastify.getSchemas() // will returns
  {
    A: { $id: 'A', properties: { hello: '...' } },
    B: { $id: 'B', properties: { hello: '...' } }
  }
```

Could you be clearer than me to write this behaviour? 😅

Reviewer: ```suggestion
`fastify.getSchemas()`, returns a hash of all schemas added via `.addSchema`. The keys of the hash are the `$id`s of the JSON Schema provided.
```

})
```

<a name="set-validator-compiler"></a>
#### setValidatorCompiler

> Set the schema validator compiler for all routes [here](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#schema-validator).

Reviewer: ```suggestion
Set the schema validator compiler for all routes. See [#schema-validator](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#schema-validator).
```

<a name="set-serializer-resolver"></a>
#### setSerializerCompiler

> Set the schema serializer compiler for all routes [here](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#schema-serializer).

Reviewer: ```suggestion
Set the schema serializer compiler for all routes. See [#schema-serializer](https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#schema-serializer).
```

<a name="validator-compiler"></a>
#### validatorCompiler

> This property can be used to get the schema validator. If not set, it will be `null` until the server starts.

Reviewer: What will it be after the server starts?

Author: It will be a function that, given a schema and the route's coordinates, return a compiled function (from ajv)

It could be useful to know if the user wants to use the default compiler (with all the shared schema) to generate some validations.
I used it [here](https://github.com/Eomm/fastify-schema-constraint/) for example, I will add this information 👍

Reviewer: My point is that you should explain this in the documentation.

<a name="serializer-compiler"></a>
#### serializerCompiler

> This property can be used to get the schema serializer. If not set, it will be `null` until the server starts.

Reviewer: What will it be after the server starts?

Author: As above: I will add this info 👍
